lib/accesslists/hierarchy.go

```diff
@@ -0,0 +1,837 @@
+	// MemberParents are AccessLists that have this list as a member.
+	MemberParents map[string]*HierarchyNode
+	// OwnerParents are AccessLists that have this list as an owner.
+	OwnerParents map[string]*HierarchyNode
+}
+
+type hierarchy struct {
+	Nodes map[string]*HierarchyNode
+	Locks services.LockGetter
+	Clock clockwork.Clock
+}
+
+type Hierarchy interface {
+	// ValidateAccessListMember validates the addition of an AccessListMember to an existing AccessList.
+	ValidateAccessListMember(parentListName string, member *accesslist.AccessListMember) error
+	// ValidateAccessListOwner validates the addition of an existing AccessList as an Owner to another existing AccessList.
+	ValidateAccessListOwner(parentListName string, owner *accesslist.Owner) error
+	// ValidateAccessListWithMembers validates the addition of a new or existing AccessList with a list of AccessListMembers.
+	ValidateAccessListWithMembers(accessList *accesslist.AccessList, members []*accesslist.AccessListMember) error
+	// GetOwners returns a flattened list of Owners for an Access List, including inherited Owners.
+	//
+	// Returned Owners are not validated for expiration or other requirements – use IsAccessListOwner
+	// to validate an Owner's ownership status.
+	GetOwners(accessListName string) ([]*accesslist.Owner, error)
+	// IsAccessListOwner determines if a User is a valid Owner of an existing or new AccessList,
+	// including via inheritance. If User has any inForce Locks, it will return an error.
+	IsAccessListOwner(ctx context.Context, user types.User, accessListName string) (MembershipOrOwnershipType, error)
+	// IsAccessListMember determines if a User is a valid Member of an existing AccessList,
+	// including via inheritance. If User has any inForce Locks, it will return an error.
+	IsAccessListMember(ctx context.Context, user types.User, accessListName string) (MembershipOrOwnershipType, error)
+	// GetInheritedGrants returns the combined Grants for an Access List's members, inherited from any ancestor lists.
+	GetInheritedGrants(accessListName string) (*accesslist.Grants, error)
+	// GetMemberParents returns Access Lists where the given Access List is a member.
+	GetMemberParents(accessListName string) ([]*accesslist.AccessList, error)
+	// GetOwnerParents returns Access Lists where the given Access List is an owner.
+	GetOwnerParents(accessListName string) ([]*accesslist.AccessList, error)
+	// GetMembers recursively fetches all non-list members for an AccessList.
+	//
+	// Returned Members are not validated for expiration or other requirements - use IsAccessListMember
+	// to validate a Member's membership status.
+	GetMembers(accessListName string) ([]*accesslist.AccessListMember, error)
+}
+
+type HierarchyConfig struct {
+	AccessLists []*accesslist.AccessList
+	Members     MembersAndLocksGetter
+	Locks       services.LockGetter
+	Clock       clockwork.Clock
+}
+
+func checkAndSetDefaults(cfg *HierarchyConfig) error {
+	if cfg.AccessLists == nil {
```

Comment:
Is empty hierarchy valid? Should probably be, just wondering.

Reply:
Hm, yeah, it should be, e.g. if you're creating your first Access List. Not sure why I spaced on that.

Reply:
Although, does Go treat an empty slice as `nil`? I assume if you simply have no existing lists, calling `cache.GetAccessLists(ctx)` wouldn't return `nil`, but I could be wrong...